### PR TITLE
LDAP support

### PR DESCRIPTION
**Pull request by thisisdarshan**
Implementing LDAP Support.
Rebase @lindskogen commit with master.

### CODE REVIEW

**Comment by MaxLeiter** [comment]
There are some style errors causing the checks to fail: https://travis-ci.org/thelounge/lounge/jobs/143304208#L160

**Comment by MaxLeiter** [comment]
Please squash your PRs [1] so we can keep a nice and concise git commit log. Other then that looks good to me :)

[1] http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

**Comment by maxpoulin64** [comment]
Agreed for the squash. I'll try to give it a try and review later in the day. We've been a little bit behind PR's lately, sorry about the delays :/ This one proved kinda hard to test as nobody on the team knows LDAP, but with the resources shared on #170 it should be a lot easier. Thanks a lot for updating that old PR!

**Comment by astorije** [comment]
This is a bit of a silly comment, but on top of the squash, something that bothers me a bit is the authorship change here. This is not a rebase, it's essentially a copy and paste, therefore completely shadowing @lindskogen's work/contribution.
It would be great if you could keep the original commit and apply your changes to it (rebase) so it appears as a commit authored by @lindskogen and rebased by you, that would in the end appear like this:
screen shot 2016-07-09 at 20 17 34
screen shot 2016-07-09 at 20 17 42

**Comment by thisisdarshan** [comment]
I have squashed my commits into one, but I am not sure how would I be able to keep the original commit.

**Comment by MaxLeiter** [comment]
https://stackoverflow.com/questions/3042437/change-commit-author-at-one-specific-commit Might be of help

**Comment by thisisdarshan** [comment]
This is all set, changed the author to @lindskogen.

**Comment by astorije** [comment]
Very nice of you, @thisisdarshan, thanks!
@lindskogen, are you OK with closing your PR in favor of this one?

**Comment by lindskogen** [comment]
Go ahead! 👍

File: src/server.js [hunk 1]
```diff
@@ -209,10 +223,38 @@ function reverseDnsLookup(socket, client) {
...
+			console.error("User (" + user + ") with no local password set tried signed in. (Probably a ldap user)");
+		}
+		result = false;
+	} finally {
+		callback(result);
+	}
+}
+
+function ldapAuth(client, user, password, callback) {
+	var bindDN = Helper.config.ldap.primaryKey + "=" + user + "," + Helper.config.ldap.baseDN;
```
Comment (maxpoulin64) [inline src/server.js]: I don't speak LDAP, but wouldn't the user here need to be escaped, for example if `user` here is `Max-P,dn=evil.max-p.me\0`?

Comment (thisisdarshan) [inline src/server.js]: @lindskogen can you comment on this?

Comment (lindskogen) [inline src/server.js]: Yeah, I guess that could be a problem

**Comment by squidboylan** [comment]
What is the status of this? We're hoping to use this soon

**Comment by thisisdarshan** [comment]
@maxpoulin64 i have escaped the user - this will address your concern.

**Comment by maxpoulin64** [comment]
Took a lot of work, but I think I finally got the thing to work! 🎉 👍

**Comment by lindskogen** [comment]
@thisisdarshan, @maxpoulin64 nice job guys! Is is ready for merge? 👍

**Comment by jedahan** [comment]
Tested on our corporate ldap, works like a dream!

**Comment by astorije** [comment]
Alright, I couldn't test this but @maxpoulin64 did plus apparently several folks here. I don't see anything that bugs me code-wise so... it's happeniiiing 🎉 
👍 and merging.

**Pull request merged**